ui/apps/platform/src/Containers/Clusters/ClustersTable.tsx

```diff
@@ -0,0 +1,246 @@
+    toggleCluster: (clusterId) => void;
+};
+
+export const COL = {
+    STATUS: 'status',
+    SENSOR: 'sensor',
+} as const;
+type ColId = (typeof COL)[keyof typeof COL];
+
+type ExpansionMap = Record<string, ColId | null>;
+
+function ClustersTable({
+    clusterIdToRetentionInfo,
+    tableState,
+    selectedClusterIds,
+    onClearFilters,
+    onDeleteCluster,
+    toggleAllClusters,
+    toggleCluster,
+}: ClustersTableProps) {
+    const metadata = useMetadata();
+    const [expanded, setExpanded] = useState<ExpansionMap>({});
+
+    function toggle(clusterId: string, col: ColId) {
+        setExpanded((prev) => ({
+            ...prev,
+            [clusterId]: prev[clusterId] === col ? null : col,
+        }));
+    }
+
+    function isCellExpanded(clusterId: string, col: ColId) {
+        return expanded[clusterId] === col;
+    }
+
+    function isRowExpanded(clusterId: string) {
+        return expanded[clusterId] != null;
+    }
+
+    function isHelmManaged(cluster: Cluster) {
+        return (
+            cluster.managedBy === 'MANAGER_TYPE_HELM_CHART' ||
+            (cluster.managedBy === 'MANAGER_TYPE_UNKNOWN' && !!cluster.helmConfig)
+        );
+    }
+
+    function isOperatorManaged(cluster: Cluster) {
+        return cluster.managedBy === 'MANAGER_TYPE_KUBERNETES_OPERATOR';
+    }
+
+    const colSpan = 9;
+
+    return (
+        <Table>
+            <Thead>
+                <Tr>
+                    <Th
+                        select={{
+                            onSelect: () => toggleAllClusters(),
+                            isSelected:
+                                tableState.type === 'COMPLETE' &&
+                                tableState.data.length === selectedClusterIds.length,
+                        }}
+                    />
+                    <Th>Cluster</Th>
+                    <Th>Provider (Region)</Th>
+                    <Th>Status</Th>
+                    <Th>Sensor upgrade status</Th>
+                    <Th>Credential expiration</Th>
+                    <Th>Cluster deletion</Th>
+                </Tr>
+            </Thead>
+            <TbodyUnified
+                tableState={tableState}
+                colSpan={colSpan}
+                errorProps={{
+                    title: 'There was an error loading cluster information',
+                }}
+                filteredEmptyProps={{ onClearFilters }}
+                renderer={({ data }) => (
+                    <>
+                        {data.map((clusterInfo, rowIndex) => {
+                            const provider = formatCloudProvider(
+                                clusterInfo.status?.providerMetadata
+                            );
+                            const clusterId = clusterInfo.id;
+                            return (
+                                <Tbody isExpanded={isRowExpanded(clusterId)} key={clusterInfo.id}>
+                                    <Tr key={clusterInfo.id}>
```

Comment:
Variation between `clusterId` and `cluserInfo.id` in these lines brought a different question to mind:

Is `key` prop of `Tr` child element redundant with `key` prop of `Tbody` parent element, which indeed needs it because of `map` method.

Reply:
Good catch, I pulled the `Tbody` into the loop after having the add the expandable sections but never removed the `Tr` key.